scheduler/src/cook/cache.clj

```diff
@@ -0,0 +1,63 @@
+(ns cook.cache
+  (:require [clj-time.core :as t])
+  (:import (com.google.common.cache Cache)))
+
+
+(defn lookup-cache!
+  "Generic cache. Caches under a key (extracted from the item with extract-key-fn. Uses miss-fn to fill
+  any misses. Caches only positive hits where both functions return non-nil"
+  [^Cache cache extract-key-fn miss-fn item]
+  (if-let [key (extract-key-fn item)]
+    (if-let [result (.getIfPresent cache key)]
+      result ; we got a hit.
+      (let [new-result (miss-fn item)]
+        ; Only cache non-nil
+        (when new-result
+          (.put cache key new-result))
+        new-result))
+    (miss-fn item)))
+
+(defn expire-key!
```

Comment:
Maybe. It could help. However, does Caffeine commit to doing expiration exactly as scheduled? 
"Expiration is performed with periodic maintenance during writes and occasionally reads"

It could reduce the issue discussed below where we might keep an 'expired' entry instead of a good entry if we hit the capacity limit.

Reply:
Expiration is exact by the timestamp, but evicting and notifying are not immediate. That would require a scheduling thread, which is a heavyweight requirement for a cache. Since it’s a cache rather than a scheduler, it’s usually not a concern. The expiration is ordered on a timerwheel for O(1) operations and periodically checked when enough activity on the cache accrues. The expired items are removed prior to size eviction, rather than act as dead weight.